{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1998015-bf26-4ccb-9ee5-c3d8ab74cf1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Premièrement, on va commencer par extract les génomes à leur bonne place.\n",
    "\n",
    "#On va set up le dictionary des cas genes qu'on veut:\n",
    "bacteria_syntenies = {}\n",
    "\n",
    "bacteria_syntenies[\"Thioalkalivibrio\"] = [\"Thioalkalivibrio_TypeIV\", \"Thioalkalivibrio_TypeI-F\"]\n",
    "bacteria_syntenies[\"ClostridiumKluyveri\"] = [\"ClostridiumKluyveri_TypeI-B\"]\n",
    "bacteria_syntenies[\"Halodurans\"] = [\"Halodurans_TypeIII-B\", \"Halodurans_TypeI-C\"]\n",
    "bacteria_syntenies[\"Staph_Epidermidis\"] = [\"Staph_Epidermidis_TypeIII-A\"]\n",
    "bacteria_syntenies[\"Rhodococcus\"] = [\"Rhodococcus_TypeIV-1\", \"Rhodococcus_TypeIV-2\"]\n",
    "bacteria_syntenies[\"Cyanothece\"] = [\"Cyanothece_TypeI-D\", \"Cyanothece_TypeI-A\"]\n",
    "bacteria_syntenies[\"Synechocystis\"] = [\"Synechocystis_TypeIII-B\", \"Synechocystis_TypeI-D\"]\n",
    "bacteria_syntenies[\"Thermotoga\"] = [\"Thermotoga_TypeIII-C\", \"Thermotoga_TypeI-B\"]\n",
    "bacteria_syntenies[\"Candidatus\"] = [\"Candidatus_TypeIII-A\", \"Candidatus_TypeI-B\"]\n",
    "bacteria_syntenies[\"Geobacter\"] = [\"Geobacter_TypeI-G\", \"Geobacter_TypeI-E\"]\n",
    "bacteria_syntenies[\"Shewanella\"] = [\"Shewanella_TypeI-F\"]\n",
    "bacteria_syntenies[\"Vibrio\"] = [\"Vibrio_TypeI-F\"]\n",
    "bacteria_syntenies[\"Yersinia\"] = [\"Yersinia_TypeI-F\"]\n",
    "bacteria_syntenies[\"Ecoli\"] = [\"Ecoli_TypeI-E\"]\n",
    "\n",
    "plasmides = {}\n",
    "#Pour pouvoir savoir quel élément du fasta file on doit checker.\n",
    "plasmides[\"Thioalkalivibrio_TypeIV\"] = 1\n",
    "plasmides[\"Rhodococcus_TypeIV-1\"] = 3\n",
    "plasmides[\"Rhodococcus_TypeIV-2\"] = 2\n",
    "plasmides[\"Synechocystis_TypeIII-B\"] = 1\n",
    "plasmides[\"Synechocystis_TypeI-D\"] = 1\n",
    "plasmides[\"Vibrio_TypeI-F\"] = 4\n",
    "\n",
    "\n",
    "\n",
    "extraction_dict = {}\n",
    "\n",
    "extraction_dict[\"Thioalkalivibrio_TypeIV\"] = {#pTK9001 Plasmid\n",
    "    \"Csf4\": (61135, 2300, 2),\n",
    "    \"Csf5\": (63526, 701, 1),\n",
    "    \"Csf1\": (64224, 758, 1),\n",
    "    \"Csf2\": (65007, 1049, 1),\n",
    "    \"Csf3\": (66038, 674, 1)\n",
    "}\n",
    "extraction_dict[\"Thioalkalivibrio_TypeI-F\"] = { #Chromosomal\n",
    "    \"Cas1\": (2566882, 977, 1),\n",
    "    \"Cas3_Cas2\": (2567856,3368, 1),\n",
    "    \"Cas6\": (2574607, 560, 1),\n",
    "    \"Csy1\": (2571311, 1274, 1),\n",
    "    \"Csy2\": (2572578, 968, 1),\n",
    "    \"Csy3\": (2573563, 1040, 1)\n",
    "}\n",
    "extraction_dict[\"ClostridiumKluyveri_TypeI-B\"] = {#Chromosomal\n",
    "    \"Cas1\": (2826093, 1004, 2),\n",
    "    \"Cas2\": (2825794, 284, 2),\n",
    "    \"Cas3\": (2827878, 2582, 2),\n",
    "    \"Cas4\": (2827124, 494, 2),\n",
    "    \"Cas5\": (2830479, 740, 2),\n",
    "    \"Cas6\": (2833938, 710, 2),\n",
    "    \"Cas7\": (2831222, 929, 2)\n",
    "} \n",
    "extraction_dict[\"Halodurans_TypeIII-B\"] = {#Chromosomal\n",
    "    \"Cmr6\": (353156, 746, 1),\n",
    "    \"Cmr5\": (352782, 377, 1),\n",
    "    \"Cmr4\": (351874, 911, 1),\n",
    "    \"Cmr3\": (350615, 1259, 1),\n",
    "    \"Cmr1\": (347483, 1376, 1),\n",
    "    \"Cas10\": (348829, 1814, 1),   \n",
    "    \"Cas3\": (358059, 2402, 1)\n",
    "}\n",
    "extraction_dict[\"Halodurans_TypeI-C\"] = {#Chromosomal\n",
    "    \"Cas1\": (364593, 1031, 1),\n",
    "    \"Cas2\": (365633, 290, 1),\n",
    "    \"Cas4\": (363937, 659, 1),\n",
    "    \"Cas5\": (360504, 710, 1),\n",
    "    \"Cas8\": (361216, 1883, 1),\n",
    "    \"Cas7\": (363096, 851, 1)\n",
    "}  \n",
    "extraction_dict[\"Staph_Epidermidis_TypeIII-A\"] = {#Chromosomal\n",
    "    \"Cas6\": (959724, 734, 2),\n",
    "    \"Csm6\": (960455, 1268, 2),\n",
    "    \"Csm3.D\": (961723, 1022, 2),\n",
    "    \"Csm4\": (962748, 908, 2),\n",
    "    \"Csm3.A\": (963667, 644, 2),\n",
    "    \"Csm2\": (964313, 425, 2),\n",
    "    \"Cas10\": (964741, 2273, 2),   \n",
    "    \"Cas1\": (967332, 905, 2),\n",
    "    \"Cas2\": (967027, 305, 2)\n",
    "}\n",
    "extraction_dict[\"Rhodococcus_TypeIV-1\"] = {#pRLH3 Plasmid\n",
    "    \"Csf4\": (6159, 440, 2),\n",
    "    \"Csf1\": (7090, 659, 2),\n",
    "    \"Csf2\": (5311, 851, 2),\n",
    "    \"Csf3\": (4577, 737, 2)\n",
    "}\n",
    "extraction_dict[\"Rhodococcus_TypeIV-2\"] = {#pRLH2 Plasmid\n",
    "    \"Csf4\": (71270, 416, 1),\n",
    "    \"Csf1\": (70596, 677, 1),\n",
    "    \"Csf2\": (71686, 866, 1),\n",
    "    \"Csf3\": (72549, 725, 1)\n",
    "}\n",
    "extraction_dict[\"Cyanothece_TypeI-D\"] = {  #Chromosomal\n",
    "    \"Cas10\": (519761, 2924, 2),\n",
    "    \"Cas1\": (515195, 977, 2),\n",
    "    \"Cas2\": (514905, 293, 2),\n",
    "    \"Cas3\": (522743, 2105, 2),\n",
    "    \"Cas4\": (516205, 590, 2),\n",
    "    \"Cas6\": (516921, 809, 2), \n",
    "    \"Csc1\": (517699, 776, 2),\n",
    "    \"Csc2\": (518618, 1070, 2)\n",
    "}\n",
    "extraction_dict[\"Cyanothece_TypeI-A\"] = {  #Chromosomal\n",
    "    \"Cas5\": (1732461, 638, 1),\n",
    "    \"Cas6\": (1729303, 665, 1),\n",
    "    \"Cas7\": (1731493, 971, 1),\n",
    "    \"Cas8\": (1729965, 1547, 1)\n",
    "}\n",
    "extraction_dict[\"Synechocystis_TypeIII-B\"] = {#pSYSA Plasmid\n",
    "    \"Cmr6\": (80203, 1937, 2),\n",
    "    \"Cmr5\": (82196, 395, 2),\n",
    "    \"Cmr4\": (82630, 779, 2),\n",
    "    \"Cmr3\": (83905, 1112, 2),\n",
    "    \"Cas10\": (85199, 2939, 2),   \n",
    "    \"Cas1\": (88627, 989, 1),\n",
    "    \"Cas2\": (89620, 278, 1)\n",
    "}\n",
    "extraction_dict[\"Synechocystis_TypeI-D\"] = {#pSYSA Plasmid\n",
    "    \"Cas10\": (8531, 2927, 1),\n",
    "    \"Cas1\": (14804, 977, 1),\n",
    "    \"Cas2\": (15813, 284, 1),\n",
    "    \"Cas3\": (6349, 2168, 1),\n",
    "    \"Cas4\": (14222, 572, 1),\n",
    "    \"Cas6\": (13441, 788, 1), \n",
    "    \"Csc1\": (12674, 764, 1),\n",
    "    \"Csc2\": (11524, 989, 1)\n",
    "}\n",
    "extraction_dict[\"Thermotoga_TypeIII-C\"] = { #Chromosomal\n",
    "    \"Cmr6\": (1254052, 848, 1),\n",
    "    \"Cmr5\": (1251786, 368, 1),\n",
    "    \"Csm3\": (1253055, 1007, 1),\n",
    "    \"Cmr3\": (1252151, 911, 1),\n",
    "    \"Cas10\": (1255658, 3656, 1)\n",
    "}\n",
    "extraction_dict[\"Thermotoga_TypeI-B\"] = {#Chromosomal\n",
    "    \"Cas1\": (214628, 983, 2),\n",
    "    \"Cas2\": (214358, 224, 2),\n",
    "    \"Cas3\": (208843, 2234, 2),\n",
    "    \"Cas4\": (215622, 491, 2),\n",
    "    \"Cas5\": (211062, 692, 2),\n",
    "    \"Cas6\": (216130, 761, 2),\n",
    "    \"Cas7\": (211751, 884, 2),\n",
    "    \"Cas8\": (212619, 1727, 2)\n",
    "}\n",
    "extraction_dict[\"Candidatus_TypeIII-A\"] = {#Chromosomal\n",
    "    \"Cas6\": (2044971, 962, 1),\n",
    "    \"Csm5\": (2051954, 1169, 1),\n",
    "    \"Csm3\": (2050162, 797, 1),\n",
    "    \"Csm4\": (2050969, 974, 1),\n",
    "    \"Csm2\": (2049709, 434, 1),\n",
    "    \"Cas10\": (2047189, 2498, 1)\n",
    "}\n",
    "extraction_dict[\"Candidatus_TypeI-B\"] = {#Chromosomal\n",
    "    \"Cas1\": (1184251, 995, 2),\n",
    "    \"Cas2\": (1183985, 266, 2),\n",
    "    \"Cas3\": (1185754, 2255, 2),\n",
    "    \"Cas4\": (1185254, 437, 2),\n",
    "    \"Cas5\": (1187996, 875, 2),\n",
    "    \"Cas6\": (1192411, 839, 2),\n",
    "    \"Cas7\": (1188921, 953, 2)\n",
    "}\n",
    "extraction_dict[\"Geobacter_TypeI-G\"] = {#Chromosomal\n",
    "    \"Cas1\": (72629, 1679, 1),\n",
    "    \"Cas2\": (74313, 287, 1),\n",
    "    \"Cas3\": (65569, 2765, 1)\n",
    "}\n",
    "extraction_dict[\"Geobacter_TypeI-E\"] = {#Chromosomal\n",
    "    \"Cas1\": (1520566, 920, 1),\n",
    "    \"Cas2\": (1521467, 305, 1),\n",
    "    \"Cas3\": (1512321, 2651, 1),\n",
    "    \"Cas5\": (1518041, 680, 1),\n",
    "    \"Cas6\": (1518708, 608, 1),\n",
    "    \"Cas7\": (1516944, 1094, 1),\n",
    "    \"Cse1\": (1514969, 1520, 1),\n",
    "    \"Cse2\": (1516473, 443, 1)\n",
    "}\n",
    "extraction_dict[\"Shewanella_TypeI-F\"] = {#Chromosomal\n",
    "    \"Cas1\": (596440, 977, 2),\n",
    "    \"Cas3_Cas2\": (593014, 3413, 2),\n",
    "    \"Csy1\": (591748, 1010, 2),\n",
    "    \"Csy2\": (590514, 950, 2),\n",
    "    \"Csy3\": (589461, 560, 2)\n",
    "}\n",
    "extraction_dict[\"Vibrio_TypeI-F\"] = {#pMLB287\n",
    "    \"Cas6\": (159413, 638, 2),\n",
    "    \"Cas3_Cas2\": (163265, 3386, 2),\n",
    "    \"Csy1\": (161994, 1271, 2),\n",
    "    \"Csy2\": (161104, 893, 2),\n",
    "    \"Cas1\": (166648, 977, 2),\n",
    "    \"Csy3\": (160061, 1031, 2)\n",
    "}\n",
    "extraction_dict[\"Yersinia_TypeI-F\"] = {#Chromosomal\n",
    "    \"Cas1\": (4676421, 980, 1),\n",
    "    \"Cas3_Cas2\": (4677398, 3290, 1),\n",
    "    \"Csy1\": (4680844, 1346, 1),\n",
    "    \"Csy2\": (4682187, 950, 1),\n",
    "    \"Cas6\": (4684170, 554, 1),\n",
    "    \"Csy3\": (4683155, 1004, 1)\n",
    "}\n",
    "extraction_dict[\"Ecoli_TypeI-E\"] = {#Chromosomal\n",
    "    \"Cas1\": (2862948, 917, 2),\n",
    "    \"Cas2\": (2862662, 284, 2),\n",
    "    \"Cas3\": (2868646, 2666, 2),\n",
    "    \"Cas5\": (2864467, 674, 2),\n",
    "    \"Cas7\": (2865144, 1091, 2),\n",
    "    \"Cse2\": (2866248, 482, 2),\n",
    "    \"Cas6\": (2863881, 599, 2),\n",
    "    \"Cse1\": (2866723, 1508, 2)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2fbb2cfa-47a7-4432-897b-213994f8d87a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/seddik/anaconda3/envs/bioinfo/lib/python3.12/site-packages (1.83)\n",
      "Requirement already satisfied: numpy in /home/seddik/anaconda3/envs/bioinfo/lib/python3.12/site-packages (from biopython) (1.26.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2d6d7147-128b-4340-bb26-0be7eb91c15a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ed9cc3a-b0b6-4db2-8147-6f19233a52a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(seq):\n",
    "    return str(Seq(seq).reverse_complement())\n",
    "\n",
    "#Directory qui contient les génomes des bactéries\n",
    "base_directory = \"/home/seddik/Documents/Synesth/iter4/Bacteria\"\n",
    "#On va utiliser ces noms pour join et trouver le actual bon path\n",
    "bacteria_names = [\"Thioalkalivibrio\", \"ClostridiumKluyveri\", \"Halodurans\", \"Staph_Epidermidis\", \"Rhodococcus\", \"Cyanothece\", \"Synechocystis\", \"Thermotoga\", \"Candidatus\", \"Geobacter\", \"Shewanella\", \"Vibrio\", \"Yersinia\", \"Ecoli\"]\n",
    "\n",
    "\n",
    "#Permet d'extraire le bon gène sous le bon format pour une certaine séquence. \n",
    "def extract_gene_sequence(record, start, length, orientation):\n",
    "    sequence = record.seq[(start-1):(start-1+length)]\n",
    "    if orientation == 2:\n",
    "        sequence = reverse_complement(sequence)\n",
    "    return str(sequence)\n",
    "    \n",
    "gene_sequences = {}\n",
    "\n",
    "#On itère à travers les directories \n",
    "for name in bacteria_names:\n",
    "        \n",
    "    bacteria_dir = os.path.join(base_directory, name)\n",
    "\n",
    "    #on a un seul file par dossier. Donc cette liste contient un seul élément.\n",
    "    fasta_files = [f for f in os.listdir(bacteria_dir)]\n",
    "    \n",
    "    #error catching\n",
    "    if not fasta_files:\n",
    "        print(f\"No FASTA file found in directory: {bacteria_dir}\")\n",
    "        continue\n",
    "    \n",
    "    #On obtient le actual absolute path de chaque GCA file (contenant le complete génome)\n",
    "    fasta_file = os.path.join(bacteria_dir, fasta_files[0])\n",
    "\n",
    "    # Initialise le dictionnaire pour la bactérie considérée\n",
    "\n",
    "    for synteny in bacteria_syntenies[name]:\n",
    "\n",
    "        gene_sequences[synteny] = {}\n",
    "\n",
    "    #Nous permet de connaitre l'indice du record qu'il faut checker dans le génome.\n",
    "    #Si le bacteria name n'est pas dans le dictionnaire des plasmides, alors on sait\n",
    "    #que le record voulu est placé dans le chromosome et donc la valeur de 0 par défaut\n",
    "    #sera donnée à l'indice. \n",
    "        record_index = plasmides.get(synteny, 0)\n",
    "\n",
    "        for i, record in enumerate(SeqIO.parse(fasta_file, \"fasta\")):\n",
    "            if i == record_index:\n",
    "            # on a le bon indice\n",
    "                for cas_gene, (start, length, orientation) in extraction_dict.get(synteny, {}).items():\n",
    "                    gene_sequence = extract_gene_sequence(record, start, length, orientation)\n",
    "                    gene_sequences[synteny][cas_gene] = gene_sequence\n",
    "                break  # Exit the loop once the desired record is processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bf8b6a1a-f922-4c63-ab61-94d5a721d5d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "output_directory = \"/home/seddik/Documents/Synesth/iter6\"\n",
    "\n",
    "for bacterium, genes in gene_sequences.items():\n",
    "    for cas_gene, sequence in genes.items():\n",
    "\n",
    "        if not sequence.strip():\n",
    "            continue\n",
    "\n",
    "        cleaned_sequence = ''.join(sequence.split())\n",
    "                                   \n",
    "        fasta_file = os.path.join(output_directory, f\"all_{cas_gene}.fna\")\n",
    "        seq_record = SeqRecord(Seq(sequence), id=bacterium, description=\"\")\n",
    "        \n",
    "        with open(fasta_file, \"a\") as output_handle:\n",
    "            SeqIO.write(seq_record, output_handle, \"fasta\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
